src/BuildScriptGenerator/Node/NodePlatform.cs

```diff
@@ -187,11 +187,13 @@ public PlatformDetectorResult Detect(RepositoryContext context)
                 throw new InvalidUsageException("Multiple monorepo package management tools are found, please choose to use either Lerna or Lage.");
             }
 
+            string yarnVersionSpec = null;
             if (ctx.SourceRepo.FileExists(NodeConstants.YarnLockFileName) || packageJson?.engines?.yarn != null)
             {
                 packageManagerCmd = NodeConstants.YarnCommand;
                 configureYarnCache = false;
                 packageInstallerVersionCommand = NodeConstants.YarnVersionCommand;
+                yarnVersionSpec = ctx.SourceRepo.FileExists(NodeConstants.YarnLockFileName) ? "latest" : packageJson?.engines?.yarn?.Value as string;
```

Comment:
Let's just stick with using the value of `engines.yarn` for the version of `yarn` we update to -- since the existence of a `yarn.lock` file is how we determine that `yarn` should be used most of the time, I'm hesitant to have us snap to a dynamic version with `latest` instead of a fixed version. There could be a breaking change introduced in the `latest` version that we're not aware of, and customers are stuck using it unless they update their `engines.yarn` field to snap to something different. I think the proper flow is to use the fixed version that we pre-bake into build images and if the user wants something different, they can specify it in their `engines.yarn` field.

Also, the logic currently has the existence of the `yarn.lock` file take precedence over the `engines.yarn` value if the user has set it, so even if the user wanted to change to something that isn't `latest`, they would be locked to it unless they got rid of their `yarn.lock` file, which wouldn't be feasible for most projects.

Reply:
Thank Cormac for the feedback. I have the same concerns that you mentioned for the `yarn` version. 
However, according to the doc -  https://github.com/microsoft/Oryx/blob/main/doc/runtimes/nodejs.md#package-manager, 
`If a yarn.lock file is found in your repo root or if you specify "yarn" in the engines field of package.json, the latest or specified version of yarn will be used instead of npm.`. 
it says to use the `latest` version of yarn if there is `yarn.lock` file is found or take the specified version of yarn if mentioned in the `engines` field of `package.json` file. By this doc, I was convinced to use the latest version of `yarn` if there is `yarn.lock` found and even take the precedence over the yarn version mentioned in the `engines` field of `package.json` file. 
Besides, there is still a scenario that can create bug. if there is `yarn.lock` file is found but no yarn version is not mentioned in the  `engines` field of `package.json` file then according to our existing code the NPM package manager would be used. According to my understanding, in this scenario the latest version of `yarn` should be used.

Based on that, is it still fine to remove the `latest` version part of yarn as you suggested?

Reply:
We should update that document to accurately reflect what the behavior is (and I would argue should be):

_If a `yarn.lock` file is found in your repo root, or if you specify a version spec for `yarn` in the `engines` field of your application's `package.json`, a stable or specified version of yarn will be used instead of npm._

This document was introduced almost four years ago, so the behavior may have changed over time without the document being updated, or the behavior was never aligned with the document. As it stands today, the document is incorrect and I think it should be updated to reflect the above behavior, which relies on us having a stable version of yarn that the customer can use, or allow them to specify a version spec that resolves to a version of yarn that works for the customer's application. I personally don't feel it's a good idea to rely on `stable` to pick the version for customers, which can introduce a breaking change at any point. If we know that the stable version we're pre-baking into our images now works, we should continue to snap to it and allow customers to update via the `engines` field if they would like to.